### PR TITLE
Parse and save to message logs csv chunk by chunk

### PR DESCRIPTION
**Pull request by jeantanzj**
## Problem
Supersedes #547 
Given a large csv file, the server can run out of memory by processing it all at once.

## Solution
- Using papaparse, chunk with manual buffer
- Takes longer but prevents oom
- Remove deduplication function since the recipients are no longer added to a set
- Removes check for invalid recipient on upload, except for Telegram
- Adds check for invalid recipient on send - for Email and SMS.

## Notes
Several tests were conducted for a node process limited to 512mb, a file of 263mb. 

|Description|Time for /upload/complete|
|--|--|
|CSVParse, in-memory, 5 requests| Crashed||
|Papaparse, chunking without added manual buffer, 1 request |54068.873ms| 
|Papaparse, chunking without added manual buffer, 5 requests |Crashed| 
|Papaparse, chunk with added manual buffer of 100mb, 1 request | 85360.497ms| 
|Papaparse, chunk with added manual buffer of 100mb, 5 request | 1110040.290ms, 118260.862ms, 147565.561ms, 150675.007ms, 149980.740ms| 
|Papaparse, chunk with added manual buffer of 10mb, 5 request |139843.858ms, 144368.612ms, 149288.904ms, 147915.836ms, 152285.827ms|
|Paparse, 15mb file,  chunk with added manual buffer of 100mb, 1 request| 8337.191ms|

### CODE REVIEW

File: backend/src/email/middlewares/email-template.middleware.ts [hunk 1]
```diff
@@ -154,42 +252,46 @@ const uploadCompleteHandler = async (
...
-      await updateCampaignAndMessages(+campaignId, s3Key, filename, records)
+    try {
+      // Continue processing
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      const transaction = await Campaign.sequelize!.transaction()
+      // - download from s3
+      const s3Client = new S3Client()
+      const downloadStream = s3Client.download(
+        //s3Key
+        '147651b0-e34c-41ce-925c-499a7e49e202' || s3Key
```
Comment (xming13) [inline backend/src/email/middlewares/email-template.middleware.ts:274]: remember to remove hard-coded key

Comment (jeantanzj) [reply backend/src/email/middlewares/email-template.middleware.ts:274]: It's a draft.

File: backend/src/core/services/parse-csv.service.ts [hunk 2]
```diff
@@ -1,65 +1,118 @@
...
+          .replace(/[\u200B-\u200D\uFEFF]/g, '')
+        return lowercaseHeader
+      },
+
+      chunk: async (rows: ParseResult<CSVParams>, parser: Papa.Parser) => {
+        parser.pause()
+        const { data, meta, errors } = rows
+        try {
+          // `rows` can have no data, but just the meta
+          if (!previewed && data.length > 0) {
```
[the remark this answers is not in the document]
Comment (jeantanzj) [reply backend/src/core/services/parse-csv.service.ts:49]: I'm not sure why, but sometimes the first chunk doesn't have data.

File: backend/src/core/services/parse-csv.service.ts [hunk 3]
```diff
@@ -1,65 +1,118 @@
...
+              )
+            }
+            await onComplete(numRecords)
+          } catch (err) {
+            reject(err)
+          }
+        }
+
+        logger.info({ message: 'Parsing complete' })
+        resolve()
```
[the remark this answers is not in the document]
Comment (jeantanzj) [reply backend/src/core/services/parse-csv.service.ts:111]: I can move the logging statement, but why?

File: backend/src/email/services/email.service.ts [hunk 4]
```diff
@@ -148,10 +158,217 @@ const getCampaignDetails = async (
...
   sendCampaignMessage,
   setCampaignCredential,
   getCampaignDetails,
   getHydratedMessage,
+  uploadCompleteOnPreview,
+  uploadCompleteOnChunk,
+  uploadCompleteOnComplete,
+  uploadProtectedCompleteOnPreview,
+  uploadProtectedCompleteOnChunk,
+  uploadProtectedCompleteOnComplete: uploadCompleteOnComplete,
```
[the remark this answers is not in the document]
Comment (jeantanzj) [reply backend/src/email/services/email.service.ts:373]: uploadProtectedComplete and uploadProtectedCompleteOnChunk both deal with EmailMessages so I thought to keep them here

File: backend/src/core/services/parse-csv.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeantanzj) [reply backend/src/core/services/parse-csv.service.ts:49]: oh sorry I misunderstood. beforeFirstChunk accepts a chunk argument, which consists some lines of the csv, unparsed. I would have to split by newlines and commas myself, which defeats the purpose of using papaparse.

**Pull request merged**